File: tests/abi/data_exchange_layout.cpp
#include "pluginterfaces/vst/ivstdataexchange.h"

#include <cstddef>
#include <cstdio>

#define PRINT_TYPE(Type) std::printf (#Type " size %zu align %zu\n", sizeof (Steinberg::Vst::Type), alignof (Steinberg::Vst::Type))
#define PRINT_OFFSET(Type, Field) std::printf (#Type "." #Field " offset %zu\n", offsetof (Steinberg::Vst::Type, Field))

template <typename T>
static void print_iid (const char* name, const T& tuid)
{
	std::printf ("%s iid", name);
	for (int index = 0; index < 16; ++index)
		std::printf (" %02X", static_cast<unsigned char> (tuid[index]));
	std::printf ("\n");
}

int main ()
{
	std::printf ("InvalidDataExchangeQueueID %u\n", Steinberg::Vst::InvalidDataExchangeQueueID);
	std::printf ("InvalidDataExchangeBlockID %u\n", Steinberg::Vst::InvalidDataExchangeBlockID);

	PRINT_TYPE (DataExchangeBlock);
	PRINT_OFFSET (DataExchangeBlock, data);
	PRINT_OFFSET (DataExchangeBlock, size);
	PRINT_OFFSET (DataExchangeBlock, blockID);

	print_iid ("IDataExchangeHandler", Steinberg::Vst::IDataExchangeHandler_iid);
	print_iid ("IDataExchangeReceiver", Steinberg::Vst::IDataExchangeReceiver_iid);
	return 0;
}

File: tools/data_exchange_layout.zig
const std = @import("std");
const data_exchange = @import("zig-vst3").pluginterfaces.vst.ivstdataexchange;

pub fn main() !void {
    const stdout = std.fs.File.stdout().deprecatedWriter();
    try stdout.print("InvalidDataExchangeQueueID {}\n", .{data_exchange.InvalidDataExchangeQueueID});
    try stdout.print("InvalidDataExchangeBlockID {}\n", .{data_exchange.InvalidDataExchangeBlockID});

    try printType(stdout, "DataExchangeBlock", data_exchange.DataExchangeBlock);
    try printOffset(stdout, "DataExchangeBlock", "data", data_exchange.DataExchangeBlock, "data");
    try printOffset(stdout, "DataExchangeBlock", "size", data_exchange.DataExchangeBlock, "size");
    try printOffset(stdout, "DataExchangeBlock", "blockID", data_exchange.DataExchangeBlock, "blockID");

    try printType(stdout, "IDataExchangeHandler", data_exchange.IDataExchangeHandler);
    try printType(stdout, "IDataExchangeReceiver", data_exchange.IDataExchangeReceiver);

    try printTuid(stdout, "IDataExchangeHandler", data_exchange.idata_exchange_handler_iid);
    try printTuid(stdout, "IDataExchangeReceiver", data_exchange.idata_exchange_receiver_iid);
}

fn printType(writer: anytype, comptime name: []const u8, comptime Type: type) !void {
    try writer.print("{s} size {} align {}\n", .{ name, @sizeOf(Type), @alignOf(Type) });
}

fn printOffset(writer: anytype, comptime type_name: []const u8, comptime field_label: []const u8, comptime Type: type, comptime field_name: []const u8) !void {
    try writer.print("{s}.{s} offset {}\n", .{ type_name, field_label, @offsetOf(Type, field_name) });
}

fn printTuid(writer: anytype, comptime name: []const u8, bytes: [16]u8) !void {
    try writer.print("{s} iid", .{name});
    for (bytes) |byte| {
        try writer.print(" {X:0>2}", .{byte});
    }
    try writer.writeByte('\n');
}
